### PR TITLE
Desktop Access File System Sharing

### PR DESCRIPTION
**Pull request by ibeckermayer**

### CODE REVIEW

**Review by zmb3 — COMMENTED** [review]
Good start! Lots of good info here, but it's almost too low-level.

Some of the sections describe specific algorithms in text, which can be hard to follow. I got about halfway through and then lost track of what is what.

When looking for feedback on a design, we want readers to consider things like:

- limitations imposed by this approach
- potential for future expansion, additional features, etc
- the "seams" where different pieces connect (the TDP<->RDP conversion, for example)

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+[Typical Operation](#typical-operation) section below. TDP won't make use of `FileId`, electing to use a file or directory's (unix-like) relative path in each
+request. The relative `path` combined with `directory_id` gives a unique "id" to any shared file-like object for any number of root-level shared directories.
+
+#### `CompletionId`
+
+`CompletionId`s are generated by the server and sent with each [`Device I/O Request`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/a087ffa8-d0d5-4874-ac7b-0494f63e2d5d).
+All [`Device I/O Request`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/a087ffa8-d0d5-4874-ac7b-0494f63e2d5d)s demand a
+[`Device I/O Response`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/1c412a84-0776-4984-b35c-3f0445fcae65) in response, and the server
+matches [`Device I/O Response`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/1c412a84-0776-4984-b35c-3f0445fcae65)s to
+[`Device I/O Request`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/a087ffa8-d0d5-4874-ac7b-0494f63e2d5d)s by the `CompletionId` field.
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply rfd/0067-desktop-access-file-system-sharing.md:76]: Sorry, yeah I know. I started doing this and then just got totally carried away... I will refactor.

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+1. server sends `IRP_MJ_CREATE` with `Path: "example.txt"`
+2. client responds with a `FileId`
+3. server sends `IRP_MJ_READ` for `FileId` for bytes 0 - 1024
+4. client executes the read and responds with the data
+5. server sends `IRP_MJ_CLOSE`
+
+## TDP File Shared Directory Extension
+
+Our RDP client lives on the Windows Desktop Service, while the directory we're sharing is exposed to us via the user's browser. This means that in order to get information to and from the shared directory,
+we must extends the TDP protocol (see the diagram in the `Introduction` section for reference).
```
Comment (zmb3) [inline rfd/0067-desktop-access-file-system-sharing.md:167]:
```suggestion
we must extend the TDP protocol (see the diagram in the `Introduction` section for reference).
```

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+the same `directory_id` is undefined behavior.
+
+`name_length` is the length in bytes of the `name`
+
+`name` is the name of the directory (without any path prefix)
+
+#### 12 - Shared Directory Acknowledge
+
+```
+| message type (12) | err error | directory_id uint32 |
```
Comment (zmb3) [inline rfd/0067-desktop-access-file-system-sharing.md:188]:
```suggestion
| message type (12) | err_code uint32 | directory_id uint32 |
```

Let's refer to this field as a `uint32` error code, not as `error` which implies some sort of more complex type.

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+```
+| message type (13) | completion_id uint32 | directory_id uint32 | path_length uint32 | path []byte |
+```
+
+This message is sent from the server to the client to request information about a file. The server will be expecting a `Shared Directory Info Response`.
+
+`completion_id` is generated by the server and must be returned by the client in its corresponding `Shared Directory Info Response` or `Shared Directory Error`
+response.
+
+`directory_id` is the `directory_id` of the top level directory being shared, as specified in a previous `Announce Shared Directory` message.
```
Comment (zmb3) [inline rfd/0067-desktop-access-file-system-sharing.md:208]:
```suggestion
`directory_id` is the `directory_id` of the top level directory being shared, as specified in a previous `Shared Directory Announce` message.
```

for consistency

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+
+`path_length` is the length in bytes of the `path`
+
+`path` is the unix-style relative path (from the root-level directory specified by `directory_id`) to the file or directory, excepting special path characters ".",
```
Comment (zmb3) [inline rfd/0067-desktop-access-file-system-sharing.md:212]: Do we have to deal with Windows-style short paths here?

https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-2000-server/cc976806(v=technet.10)

**Review by zmb3 — COMMENTED** [review]
This looks 100x better. Nice job!

The last thing I'd like us to explore before finalizing this is how RDP clients implement a read-only mode. Do we just send an error response to any mutating action, or is there something at the protocol level when you announce the directory that marks it as read-only?

File: rfd/0067-desktop-access-file-system-sharing.md (hunk 5 above)
Comment (ibeckermayer) [reply rfd/0067-desktop-access-file-system-sharing.md:212]: From my research, it seems that its no longer commonly used (i.e. [this](https://barreto.home.blog/2012/11/13/windows-server-2012-file-server-tip-disable-8-3-naming-and-strip-those-short-names-too/)). I also didn't see anything in FreeRDP that deals with it.

File: rfd/0067-desktop-access-file-system-sharing.md (hunk 2 above)
Comment (LKozlowski) [inline rfd/0067-desktop-access-file-system-sharing.md:167]: Would it be possible to add a little more clarification about what TDP is? Since we are extending the protocol maybe it would be helpful for people to add some hyperlinks to other rfd or parts in the codebase to provide context for what we're trying to achieve? Either here or in the introduction.

File: rfd/0067-desktop-access-file-system-sharing.md [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+3. server sends `IRP_MJ_READ` for `FileId` for bytes 0 - 1024
+4. client executes the read and responds with the data
+5. server sends `IRP_MJ_CLOSE`
+
+## TDP File Shared Directory Extension
+
+Our RDP client lives on the Windows Desktop Service, while the directory we're sharing is exposed to us via the user's browser. This means that in order to get information to and from the shared directory,
+we must extends the TDP protocol (see the diagram in the `Introduction` section for reference).
+
+Each `* Request` and `* Response` TDP message contains a `completion_id` field, with `* Request`s being responsible for generating the `completion_id`s, and `* Response`s being responsible for
```
Comment (LKozlowski) [inline rfd/0067-desktop-access-file-system-sharing.md:169]: Could you please tell me (and possibly describe it in this section) what is `* Request` or `* Response`? Does it mean `any` request or response? It's a bit unclear to me what does it mean in this context.

File: rfd/0067-desktop-access-file-system-sharing.md (hunk 2 above)
Comment (LKozlowski) [reply rfd/0067-desktop-access-file-system-sharing.md]: Awesome, thanks!

**Review by zmb3 — APPROVED** [review]

**Review by LKozlowski — APPROVED** [review]
Looks good! The only thing that I'm not sure about is how Windows Server will react to the access denied messages sent when we use "read only mode". It seems that it should be fine, but we never know. I think we'll have to implement it and see how it is working. I suppose when you were reading Windows' protocols designs you haven't seen any read only flags or attributes for files or directories that we can set when we share list of files so that's why we have to send these responses?

**Comment by ibeckermayer** [comment]
> > I suppose when you were reading Windows' protocols designs you haven't seen any read only flags or attributes for files or directories that we can set when we share list of files so that's why we have to send these responses?
> 
> Good point. I vaguely recall considering that, but its not clear in my mind and I may have been mistaken or lazy. I'll double check.

Double checked -- there are some RDP data structures that give us an option to respond with `FileAttributes`
https://github.com/gravitational/teleport/blob/83ca5466db8ce16568706f0cec0ccc792b15b091/lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs#L2069
which we can set to `FILE_ATTRIBUTE_READONLY`
https://github.com/gravitational/teleport/blob/83ca5466db8ce16568706f0cec0ccc792b15b091/lib/srv/desktop/rdp/rdpclient/src/rdpdr/flags.rs#L79

Probably we should do this whenever possible in read-only mode, but there are problems with relying on that. One initial one I came up with is that it's not clear what happens in every situation where a file can potentially be written to. For example, if a user creates a new file in the shared directory, are we guaranteed to have a chance to send RDP a structure with `FILE_ATTRIBUTE_READONLY` set so that it doesn't ever try to write to it? Because RDP doesn't specify behavior at this level of detail, its not clear. We could  look at FreeRDP's server implementation and/or test things empirically, but that doesn't guarantee all such implementations behave that way.

This led me to the more fundamental problem that we don't have control over the RDP server (the windows desktop itself), and so even if setting `FILE_ATTRIBUTE_READONLY` in every situation was plausible, a malicious server implementation could just choose to ignore it and write to such files anyways, breaking our security guarantee. A standard RDP client-server system doesn't face this problem, because the client is running directly on the host machine, and relaying OS-level information about the filesystem -- read-only truly means read-only, and is enforced by the OS on the client even if the server is malicious. But in our case, we're proxying everything to/from the browser's File System Access API, which doesn't give us an enforceable option like "open a directory in read only mode". Therefore regardless of how RDP works, we need to enforce read-only mode ourselves in order to guarantee it.

@LKozlowski
